nvflare/fuel/utils/config_factory.py

```diff
@@ -0,0 +1,134 @@
+
+
+class ConfigFactory:
+    logger = logging.getLogger(__qualname__)
+
+    @classmethod
+    def ordered_search_extensions(cls) -> List:
+        search_sequence: List[ConfigFormat] = ConfigFormat.search_order()
+        extensions = []
+        for fmt in search_sequence:
+            exts = ConfigFormat.get_extensions(fmt.name)
+            if exts:
+                for ext in exts:
+                    extensions.append((fmt, ext))
+
+        return extensions
+
+    @staticmethod
+    def search_config_format(init_file_path, search_dirs: Optional[List[str]] = None) \
+            -> (Optional[ConfigFormat], Optional[str]):
+        """
+        find the configuration format and the location (file_path) for given initial init_file_path and search directories.
+        for example, the initial config file path given is config_client.json
+        the search function will ignore the .json extension and search "config_client.xxx" in the given directory in
+        specified extension search order. The first found file_path will be used as configuration.
+        the ".xxx" is one of the extensions defined in the configuration format.
+
+        Args:
+            init_file_path: initial file_path for the configuration
+            search_dirs: search directory. If none, the parent directory of init_file_path will be used as search dir
+
+        Returns: Tuple of None,None or ConfigFormat and real configuration path
+
+        """
+        logger = ConfigFactory.logger
+        if not search_dirs:
+            parent_dir = pathlib.Path(init_file_path).parent
+            search_dirs = [str(parent_dir)]
+        # we ignore the original extension
+        file_basename = os.path.splitext(pathlib.Path(init_file_path).name)[0]
+        for search_dir in search_dirs:
+            logger.debug(f"search file:{file_basename} basename, search dirs = {search_dirs}")
+            for root, dirs, files in os.walk(search_dir):
+                for fmt, ext in ConfigFactory.ordered_search_extensions():
+                    found_one = ConfigFactory._search_config_file(file_basename, fmt, ext, root, files)
+                    if found_one:
+                        return found_one
+
+        return None, None
+
+    @staticmethod
+    def _search_config_file(base_filename, fmt, ext, root, files):
+        file = f"{base_filename}{ext}"
+        if file in files:
+            config_file_path = os.path.join(root, file)
+            return fmt, config_file_path
+
+        return None, None
+
+    @staticmethod
+    def has_config(init_file_path: str, search_dirs: Optional[List[str]] = None) -> bool:
+        _, real_file_path = ConfigFactory.search_config_format(init_file_path, search_dirs)
+        return real_file_path is not None
+
+    @staticmethod
+    def match_config(parent, init_file_path, match_fn) -> bool:
+        # we ignore the original extension
+        basename = os.path.splitext(init_file_path)[0]
```

Comment:
What is basename has "." in it?

Reply:
Right, splitext doesn't know ".json.default" should be treated as one ext.
I wonder maybe we should change the treatment of default files. Instead of xxx.json.default, why don't we change it to default.xxx.json? This will make it much easier for you, right?
This is totally under our control. Of course we will need to change the Provision to produce default files with the new names.

Reply:
Also xxx.json.default is anti-pattern - technically the ext here is ".default". Many tools rely on ext to work.